src/view/switch.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright 2022 The Druid Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::any::Any;
+
+use crate::view::ViewMarker;
+use crate::{view::Id, widget::ChangeFlags, MessageResult};
+
+use super::{Cx, View};
+
+pub struct Switch {
+    is_on: bool,
+}
+
+pub fn switch(is_on: bool) -> Switch {
+    Switch::new(is_on)
+}
+
+impl Switch {
+    pub fn new(is_on: bool) -> Self {
+        Switch { is_on }
+    }
+}
+
+impl ViewMarker for Switch {}
+
+impl<A> View<bool, A> for Switch {
```

Comment:
I think `bool` should actually be a generic parameter `T`, otherwise it's necessary to "`Adapt`" this everywhere, where it's used, unless the central app state is really just a `bool` :)

Reply:
Is the purpose of the callback function to be able to access specific fields from the `AppData`?

Reply:
Yes kind of.
It's mostly for mutating the actual app state, without restricting this to only accept `bool` as `AppState`. I think it's a slightly nicer form than to wrap it in an `Adapt` node, and "hiding" the mutation inside the `View` impl.

TMU the actual logic of mutating the `AppState` should reside in the `app_logic` function, so that it's obvious what happens. It's also a little bit more flexible, as you can "react" to the event directly in the callback, instead of "guessing" what has changed (in the next update). Not that it would make any realworld sense, but you could do something like this:

```rust
fn my_view(state: &mut AppState) -> impl View<AppState> {
    switch(state.my_switch_value, |state, value| {
        state.count += 1;
        if state.count > 3 {
            state.my_switch_value = value;
        }
    })
}
```